refactor datastar sdk

magiconair opened the pull request:
Refactor the Datastar TypeScript SDK according to comments in https://discord.com/channels/1296224603642925098/1301258555936215172

* Create separate types for the different messages
* Use descriptive names for the fields
* Decouple DS from hono
* Provide `writeSSE` helper method for integration with `hono/streaming` 

`id` is still optional b/c it currently breaks my code and I would like to understand where the state is kept when the client wants to pick up from a previous message. I am not sure whether that is required in all cases. 

Lets focus on whether this code is cleaner than the previous hack.

magiconair committed:
refactor datastar sdk

magiconair committed:
honor minfy default

magiconair committed:
add writeSSE helper for hono

magiconair committed:
add close option to writeSSE

magiconair committed:
handle options better, allow for viewTransitions to be undefined

viewTransitions == undefined -> datastar library default

magiconair committed:
Rename msg to event

Superpat commented:
Looks better! I've got some notes, but I'd prefer to wait untill it's been refactored to follow the [SDK](https://github.com/starfederation/datastar/blob/main/design/SDK.md). 

For the record, my main thought is I'm not sure tagged types are a necessity for the events, seeing as they all have different fields and we can differentiate them that way. But I might be wrong.

Superpat commented:
Oh wait, you have read the sdk already. There's some stuff missing though I dont have time to look into it more this morning.

magiconair commented:
> For the record, my main thought is I'm not sure tagged types are a necessity for the events, seeing as they all have different fields and we can differentiate them that way. But I might be wrong.

Might be a dumb question what do you mean with "tagged types"? I tried to understand the difference between interfaces and types in TS and found that there virtually none. (https://medium.com/@martin_hotell/interface-vs-type-alias-in-typescript-2-7-2a8f1777af4c)

I only decided to make `Event` an interface since it never gets instantiated. If you mean the `type` parameter within the types itself then I've used that both for type switching and the event type. Couldn't find a better way to switch on the types. I thought it would be awkward as well but honestly I think it makes the code more readable and easier to explain to my colleagues. Here is an example from my current project. Less magic.

```typescript
    app.get('/repo/:repo/history', async (c) => {
        return streamSSE(c, async (stream) => {
            await writeSSE(stream, noUiError());
            try {
                const resp = await blobs.listCommits({ repo: c.req.param('repo') });
                await writeSSE(stream, { type: 'signal', store: { location: `Blobs > Repos > ${resp.repo}` } });
                await writeSSE(stream, { type: 'fragment', frag: HistoryView(basePath, resp).toString() }, 'close');
            } catch (e: any) {
                await writeSSE(stream, uiError(c), 'close');
            }
        });
    });

```

Superpat commented:
A tagged type is a union of different types with one shared key that 
allows differentiating them. Types in typescript are structural, meaning 
it's not the name of the type that's important but it's shape.

It's useful when you have a huge union, but it's not the cleanest thing, 
since you have to add it everywhere after. I'll have to dive into the 
code more precisely (maybe tonight), but I think it's not necessary. At 
least, the definition I had before did not require it, but we can 
compare the two after.

magiconair committed:
rename delete to remove, add missing attributes

magiconair committed:
support stream event ids and retry

magiconair commented:
@Superpat I've added a `DatastarStream` class to support the `id` injection and retry. It doesn't seem possible to wrap the Hono streaming API class fully because it has private properties. Or I just don't know how to do that. Let me know what you think.

magiconair commented:
@Superpat I've also added the missing `settle` attribute to `fragment`, renamed `delete` to `remove` and added support for the selector and settle time to `remove` as well.
